File: algorithms/sorting/insertion_sort.cpp
#include <iostream>
#include <vector>
#include <cassert>

using namespace std;

typedef vector<int> vi;

void insertion_sort(vi &arr)
{
  for (int i = 0; i < arr.size(); i++)
  {
    while (i > 0 && arr[i - 1] >= arr[i])
    {
      int temp = arr[i];
      arr[i] = arr[i - 1];
      arr[i - 1] = temp;
      i--;
    }
  }
}

bool test_function(vector<vi> input)
{
  vi test_case = input[0];
  vi expected = input[1];
  insertion_sort(test_case);

  for (int i : test_case)
  {
    cout << i << " ";
  }

  cout << "\n";

  if (test_case == expected)
  {
    cout << "Pass"
         << "\n";

    return true;
  }
  else
    cout << "Fail"
         << "\n";

  return false;
}

int main()
{
  vector<vi> input_1 = {vi{3, 2, 1}, vi{1, 2, 3}};
  assert(test_function(input_1));

  vector<vi> input_2 = {vi{1, 2, 3}, vi{1, 2, 3}};
  assert(test_function(input_2));

  vector<vi> input_3 = {vi{}, vi{}};
  assert(test_function(input_3));

  return 0;
}

File: techniques/sorting/insertion-sort/insertion-sort.py
# SOURCE: https://www.interviewcake.com/concept/python3/insertion-sort
import random


def insertion_sort(arr):
    for i in range(len(arr)):
        while i > 0 and arr[i - 1] >= arr[i]:
            arr[i], arr[i - 1] = arr[i - 1], arr[i]
            i -= 1

    return arr


def test(input):
    test_case = input[0]
    expected = input[1]
    test_case = insertion_sort(test_case)

    print(test_case)
    if test_case == expected:
        print("Pass")
        return True
    else:
        print("Fail")

    return False


test_case_4_in = [i for i in range(1, 10)]
test_case_4_expect = [i for i in range(1, 10)]
random.shuffle(test_case_4_in)

if __name__ == "__main__":
    try:
        assert(test([[3, 2, 1], [1, 2, 3]]))
        assert(test([[], []]))
        assert(test([[1, 2, 3], [1, 2, 3]]))
        assert(test([test_case_4_in, test_case_4_expect]))
        print("All tests passed!")
    except:
        print("Wrong answer!")
